include/GafferScene/EditScopeAlgo.h

```diff
@@ -0,0 +1,83 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2019, Cinesite VFX Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFERSCENE_EDITSCOPEALGO_H
+#define GAFFERSCENE_EDITSCOPEALGO_H
+
+#include "Gaffer/EditScope.h"
+#include "Gaffer/TransformPlug.h"
+
+#include "GafferScene/ScenePlug.h"
+
+namespace GafferScene
+{
+
+namespace EditScopeAlgo
+{
+
+// Pruning
+// =======
+
+GAFFERSCENE_API void setPruned( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path, bool pruned );
+GAFFERSCENE_API void setPruned( Gaffer::EditScope *scope, const IECore::PathMatcher &paths, bool pruned );
+GAFFERSCENE_API bool getPruned( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
+
+// Transforms
+// ==========
+//
+// These methods manipulate the local transform of a location.
+
+GAFFERSCENE_API void setTransform( Gaffer::EditScope *scope, const ScenePlug::ScenePath &path, const Imath::M44f &matrix );
+GAFFERSCENE_API boost::optional<Imath::M44f> getTransform( const Gaffer::EditScope *scope, const ScenePlug::ScenePath &path );
```

Comment:
How do you feel about the potential ambiguity of keeping the name this bare? `getTransform` is somewhat ambiguous when you step back a bit (eg: it could be returning the transform for that path only if it exists at the edit scope). Would it be worth establishing a convention where we use `Edit` or similar as a suffix for these kind of things, eg: `setTransformEdit`, `getTransformEdit`, or similar? Especially as we're suffixing the processor registrations with `Edits`.

At the very least I think it'd be good to clarify the meaning of the optional in the comments.

Reply:
I first thought that having it in EditScopeAlgo was enough, but I think you're right. For me, the key realisation is that the edit is distinct from the property, and that the EditScopeAlgo API isn't suitable for querying the property (because you will usually be querying that downstream). The method names should make that clear.

But the more I think about it, the more I think `setTransformEdit/getTransformEdit` have no place in the API. They don't cope in situations where an edit may have been keyframed, or an expression stuck on. The pattern doesn't extend particularly well to TweakPlugs either. I propose a bare `has/acquire/remove` pattern, with all else done via direct interaction with plugs :

`
bool hasTransformEdit( scope, path );
struct TransformEdit; // Just the current TransformPlugs renamed
boost::optional<TransformEdit> acquireTransformEdit( scope, path, bool createIfNecessary );
void removeTransformEdit( scope, path );
`

For something like a parameter edit, we'd probably just return `TweakPlug` directly rather than another custom struct : `TransformEdit` itself would be somewhat unique to the transform case. What do you think?

Reply:
> I propose a bare has/acquire/remove pattern, with all else done via direct interaction with plugs...

Winner winner chicken dinner.

Reply:
Implemented in https://github.com/GafferHQ/gaffer/pull/3666/commits/2f26bedf1f641e188be3cfb5abc88305c1a2ad4a